creator-node/src/tracer.ts

```diff
@@ -65,6 +85,79 @@ export const setupTracing = () => {
   provider.register()
 }
 
+/**
+ * Higher-order function that adds opentelemetry tracing to a function.
+ * This wrapper works for both sync and async functions
+ *
+ * @param {string?} param.name optional name to give to the span, defaults to the function name
+ * @param {TFunction} param.fn the generic function to instrument
+ * @param {SpanOptions} param.options objects to pass into the span
+ * @returns the instrumented function
+ * @throws rethrows any errors from the original fn
+ *
+ * Usage of this would look like
+ * ```
+ * const someFunction = instrumentTracing({ fn: _someFunction })
+ * const result = someFunction(args))
+ * // or
+ * const result = await someFunction(args)
+ * ```
+ */
+export const instrumentTracing = <TFunction extends (...args: any[]) => any>({
+  name,
+  fn,
+  options
+}: {
+  name?: string
+  fn: TFunction
+  options?: SpanOptions
+}) => {
+  // build a wrapper around `fn` that accepts the same parameters and returns the same return type
+  const wrapper = (...args: Parameters<TFunction>): ReturnType<TFunction> => {
+    const spanName = name || fn.name
+    const spanOptions = options || {}
+    return tracing
+      .getTracer()
+      .startActiveSpan(spanName, spanOptions, (span: Span) => {
+        try {
+          tracing.setSpanAttribute(tracing.CODE_FUNCTION, fn.name)
+
+          // TODO add skip parameter to instrument testing function to NOT log certain args
+          // tracing.setSpanAttribute('args', JSON.stringify(args))
+          const result = fn.call(this, ...args)
+
+          // if `fn` is async, await the result
+          if (result && result.then) {
+            /**
+             * by handling promise like this, the caller to this wrapper
+             * can still use normal async/await syntax to `await` the result
+             * of this wrapper
+             * i.e. `const output = await instrumentTracing({ fn: _someFunction })(args)`
+             */
+            return result.then((val: any) => {
```

Comment:
do you have links or docs for how this works?

Reply:
I did at one point. This code block is loosely based on something I found from stackoverflow so I'll look for that. If I can't find it, I'll at least have the tests added that show that it works

Reply:
https://github.com/klny/function-wrapper/blob/master/src/wrapper.js#L25

Reply:
This isn't the stack overflow post but this is another repo doing something similar

Reply:
oh this is great. could you add a link to this file in the jsdoc above?